### PR TITLE
Add Blurhash placeholders for federated images

### PR DESCRIPTION
**Pull request by pfefferle**
## Proposed changes:

Generates [Blurhash](https://blurha.sh/) placeholders for image attachments so federated posts emit `attachment[].blurhash` — the colored-blur preview Pixelfed, Mastodon, and other fediverse clients paint while the full image loads. Without it, federated WordPress photos sit on a grey placeholder where native uploads paint instantly.

This migrates the feature from [Automattic/FOSSE](https://github.com/Automattic/fosse), **replacing its `kornrunner/blurhash` dependency with a first-party encoder** so the plugin ships Blurhash support out of the box — no new runtime dependency.

* `includes/class-blurhash-encoder.php` — a self-contained port of the public [Wolt Blurhash](https://github.com/woltapp/blurhash) encode algorithm (base83 + DCT + sRGB↔linear). MIT; header credits the Wolt algorithm and the FOSSE adaptation.
* `includes/class-blurhash.php` — orchestration (ported from FOSSE): computes the hash at upload time (cron-deferred off `wp_generate_attachment_metadata` so the upload UI isn't blocked), stores it as `_activitypub_blurhash` postmeta, and injects it into outbound Image attachments via the `activitypub_attachment` filter. GD-gated; a no-op (never blocks upload or federation) on hosts without GD or on any failure path.
* `includes/cli/class-blurhash-command.php` — `wp activitypub blurhash backfill [--dry-run] [--limit=<n>] [--force]` to hash an existing media library.

The `blurhash` JSON-LD context term already shipped in #3327; this populates it.

Notable hardening carried over from / added on top of FOSSE: stored hashes are validated against the base83 alphabet and a length cap before federating (guards postmeta poisoning), source paths are realpath-contained under the uploads dir, source size (8 MiB) and encode edge (64px) are capped, and GD image resources are freed in a `finally` block (matters for the long CLI backfill on PHP 7.4).

## Other information:

- [x] Have you written new tests for your changes, if applicable?

13 tests: encoder format + DC round-trip + edge cases, orchestration (get/set/delete, malformed-hash rejection, injection, hook registration), and a real-image end-to-end encode through GD.

## Testing instructions:

1. Upload a JPEG/PNG to the media library.
2. Run WP-Cron (or wait for it), then check the attachment: `get_post_meta( $id, '_activitypub_blurhash', true )` returns a base83 string.
3. For existing media: `wp activitypub blurhash backfill` (try `--dry-run` first).
4. Fetch a post that includes the image as an ActivityPub object and confirm its `attachment[]` entry has a `blurhash` field.
5. On a host without GD, confirm uploads and federation still work (just no `blurhash` field).

### Changelog entry

Included at `.github/changelog/add-blurhash`.

### CODE REVIEW

File: includes/class-blurhash-encoder.php [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+			for ( $x = 0; $x < $components_x; $x++ ) {
+				$normalisation = ( 0 === $x && 0 === $y ) ? 1.0 : 2.0;
+				$r             = 0.0;
+				$g             = 0.0;
+				$b             = 0.0;
+				for ( $i = 0; $i < $width; $i++ ) {
+					for ( $j = 0; $j < $height; $j++ ) {
+						$basis = $normalisation
+							* \cos( M_PI * $x * $i / $width )
+							* \cos( M_PI * $y * $j / $height );
```
[the remark this answers is not in the document]
Comment (pfefferle) [reply includes/class-blurhash-encoder.php]: Switched to `pi()` in 073070157. (`M_PI` is part of ext-standard and always defined in normal builds, but `pi()` is equivalent and removes any doubt.)

File: includes/class-blurhash.php [hunk 2]
```diff
@@ -0,0 +1,576 @@
...
+					$row[]  = array( $colors['red'], $colors['green'], $colors['blue'] );
+				}
+				$pixels[] = $row;
+			}
+
+			$hash = Blurhash_Encoder::encode( $pixels, self::COMPONENTS, self::COMPONENTS );
+			return \is_string( $hash ) && '' !== $hash ? $hash : null;
+		} catch ( \Throwable $e ) {
+			return null;
+		} finally {
```
[the remark this answers is not in the document]
Comment (pfefferle) [reply includes/class-blurhash.php:301]: Leaving as-is: `composer lint` passes on this file, so VariableAnalysis isn't flagging the unused `$e` (it allows unused caught exceptions). And on PHP 7.4 the catch variable can't be omitted — non-capturing `catch ( \Throwable )` is PHP 8.0+ — so naming it is required here.

File: tests/phpunit/tests/includes/class-test-blurhash.php [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+ * @coversDefaultClass \Activitypub\Blurhash
+ */
+class Test_Blurhash extends \WP_UnitTestCase {
+
+	/**
+	 * Clean up any scheduled cron events after each test.
+	 */
+	public function tear_down() {
+		\wp_clear_scheduled_hook( Blurhash::CRON_HOOK );
+		parent::tear_down();
```
[the remark this answers is not in the document]
Comment (pfefferle) [reply tests/phpunit/tests/includes/class-test-blurhash.php:24]: Good catch — fixed in 073070157. Switched to `wp_unschedule_hook()` so events scheduled with the attachment-ID arg are cleared between tests regardless of args.

File: includes/class-blurhash.php [hunk 4]
```diff
@@ -0,0 +1,576 @@
...
+
+		// Invalidate any prior hash so cron will re-encode against
+		// the latest bytes. `wp_generate_attachment_metadata` fires
+		// on initial upload AND on media-replace / crop / regen, so
+		// without this delete a replaced image would keep federating
+		// the placeholder for its prior bytes.
+		self::delete( $attachment_id );
+
+		self::schedule( $attachment_id );
+		return $metadata;
```
[the remark this answers is not in the document]
Comment (pfefferle) [reply includes/class-blurhash.php:410]: Good call — fixed in 529c83461. `is_encodable_attachment()` now also checks `imagetypes()` (via a new `host_can_decode()` helper, with `defined()` guards since `IMG_AVIF` is 8.1+). So on a GD build that can't decode the format, the attachment is treated as non-encodable: `schedule_encode()` bails before `delete()` (preserving any prior hash), no doomed cron is queued, and the cron handler + CLI backfill skip it cleanly — matching the existing GD-absent guard.

File: includes/class-blurhash-encoder.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (pfefferle) [reply includes/class-blurhash-encoder.php]: Already addressed in 073070157: the encoder uses `pi()` instead of `M_PI`, so there is no dependency on the constant being defined.

File: includes/class-blurhash.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (pfefferle) [reply includes/class-blurhash.php:301]: False positive here. Non-capturing `catch` is PHP 8.0+, and the project still supports PHP 7.4, where the `catch` variable is required syntax. PHPCS (with VariableAnalysis enabled) passes on this file, so we are keeping `$e`.

File: tests/phpunit/tests/includes/class-test-blurhash.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (pfefferle) [reply tests/phpunit/tests/includes/class-test-blurhash.php:24]: Already addressed in 073070157: `tear_down()` now calls `wp_unschedule_hook( Blurhash::CRON_HOOK )`, which clears the events regardless of their attachment-ID args.

File: includes/class-blurhash.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (pfefferle) [reply includes/class-blurhash.php:410]: Already addressed in 529c83461: `is_encodable_attachment()` now gates on `host_can_decode()` (via `imagetypes()` + `IMG_*` flags), and `schedule_encode()` bails before `delete()`/`schedule()` when this GD build cannot decode the format. A previously good hash is preserved and no doomed cron event is queued.

File: includes/transformer/class-post.php [hunk 5]
```diff
@@ -954,6 +954,25 @@ protected function get_media_from_blocks( $blocks, $media ) {
 						}
 					}
 					break;
+				case 'core/media-text':
+					if ( ! empty( $block['attrs']['mediaId'] ) ) {
+						// Media & Text holds either an image or a video; the default is image.
+						if ( 'video' === ( $block['attrs']['mediaType'] ?? 'image' ) ) {
+							$media['video'][] = array( 'id' => $block['attrs']['mediaId'] );
+						} else {
```
[the remark this answers is not in the document]
Comment (pfefferle) [reply includes/transformer/class-post.php:975]: Fixed in 878b91cf2: the Media & Text video branch now extracts the `<video poster>` into `icon`, matching the `core/video` handling.

File: includes/transformer/class-post.php [hunk 6]
```diff
@@ -954,6 +954,25 @@ protected function get_media_from_blocks( $blocks, $media ) {
...
+							$alt       = '';
+							$processor = new \WP_HTML_Tag_Processor( $block['innerHTML'] );
+							if ( $processor->next_tag( array( 'tag_name' => 'img' ) ) ) {
+								$alt = $processor->get_attribute( 'alt' ) ?? '';
+							}
+
+							$media['image'][] = array(
+								'id'  => $block['attrs']['mediaId'],
+								'alt' => $alt,
+							);
```
[the remark this answers is not in the document]
Comment (pfefferle) [reply includes/transformer/class-post.php]: Fixed in 878b91cf2: the Media & Text image branch now updates `alt` in place when the ID was already collected, mirroring `core/image`, so a duplicate ID no longer drops the entry or loses its alt text.

File: includes/class-blurhash.php [hunk 7]
```diff
@@ -0,0 +1,622 @@
...
+	 * @param string $hash          Encoded blurhash string.
+	 */
+	public static function set( int $attachment_id, string $hash ): void {
+		\update_post_meta( $attachment_id, self::META_KEY, $hash );
+	}
+
+	/**
+	 * Delete any stored blurhash for an attachment. Used by the
+	 * WP-CLI backfill `--force` path.
+	 *
```
[the remark this answers is not in the document]
Comment (pfefferle) [reply includes/class-blurhash.php:188]: Fixed in 878b91cf2: corrected the `delete()` docblock to state it serves the upload/regen invalidation path (`schedule_encode()`), not the CLI `--force` backfill (which encodes-then-sets).

File: tests/phpunit/tests/includes/class-test-blurhash.php [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+
+		// Build a small solid-color JPEG in the uploads dir and register it as an attachment.
+		$upload = \wp_upload_dir();
+		$file   = \trailingslashit( $upload['path'] ) . 'blurhash-test.jpg';
+		$image  = \imagecreatetruecolor( 64, 64 );
+		\imagefilledrectangle( $image, 0, 0, 63, 63, \imagecolorallocate( $image, 200, 100, 50 ) );
+		\imagejpeg( $image, $file );
+		\imagedestroy( $image );
+
+		$attachment_id = self::factory()->attachment->create_upload_object( $file );
```
[the remark this answers is not in the document]
Comment (pfefferle) [reply tests/phpunit/tests/includes/class-test-blurhash.php]: Fixed in 878b91cf2: the test now uses the committed `data/assets/sample-image.jpg` fixture, so it depends only on GD decode support, not on `imagejpeg()` write support.
